src/java/org/apache/cassandra/io/compress/ZSTDCompressor.java

```diff
@@ -0,0 +1,138 @@
+
+    public static ZSTDCompressor create(Map<String, String> compressionOptions)
+    {
+        validateCompressionLevel(parseCompressionLevelOption(compressionOptions));
+        return instance;
+    }
+
+    private static void validateCompressionLevel(int compressionLevel)
+    {
+        if (compressionLevel < FAST_COMPRESSION || compressionLevel > BEST_COMPRESSION)
+        {
+            throw new IllegalArgumentException(
+                "ZSTD compression_level " + Integer.toString(compressionLevel) + " invalid ",
+                null
+            );
+        }
+    }
+
+    private static int parseCompressionLevelOption(Map<String,String> compressionOptions)
+    {
+        return Integer.parseInt(compressionOptions.getOrDefault(COMPRESSION_LEVEL_OPTION_NAME,
+                                                                Integer.toString(DEFAULT_LEVEL)));
+    }
+
+    @Override
+    public int initialCompressedBufferLength(int chunkLength)
+    {
+        return (int)Zstd.compressBound(chunkLength);
+    }
+
+    @Override
+    public int uncompress(byte[] input, int inputOffset, int inputLength, byte[] output, int outputOffset)
+    throws IOException
+    {
+        long decompressSize = Zstd.decompressByteArray(output,
+                                                       outputOffset,
+                                                       output.length - outputOffset,
+                                                       input,
+                                                       inputOffset,
+                                                       inputLength);
+        if (Zstd.isError(decompressSize))
+        {
+            throw new IOException("ZSTD uncompress failed with error reason " + Zstd.getErrorName(decompressSize));
+        }
+
+        return (int) decompressSize;
+    }
+
+    @Override
+    public void uncompress(ByteBuffer input, ByteBuffer output) throws IOException
+    {
+        Zstd.decompress(output, input);
+    }
+
+    @Override
+    public void compress(ByteBuffer input, ByteBuffer output) throws IOException
+    {
+        if (!input.isDirect())
+        {
+            throw new IllegalArgumentException("input must be a direct buffer");
+        }
+
+        if (!output.isDirect())
+        {
+            throw new IllegalArgumentException("output must be a direct buffer");
+        }
+        long compressedSize = Zstd.compressDirectByteBuffer(output,
+                                                            output.position(),
+                                                            output.limit() - output.position(),
+                                                            input,
+                                                            input.position(),
+                                                            input.limit() - input.position(),
+                                                            compressionLevel);
```

Comment:
From what I can tell this doesn't ever change even if the level setting is changed? Am I missing something or is it effectively always 3?

Reply:
Yes, the compression level setting isn't applied